executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+                }
+                Effect::Assertion(Assertion {
+                    lhs,
+                    rhs,
+                    expected_equal,
+                }) => InterpreterAction::Assertion(
+                    var_mapper.map_expr_to_rpn(lhs),
+                    var_mapper.map_expr_to_rpn(rhs),
+                    *expected_equal,
+                ),
+                Effect::MachineCall(id, arguments) => {
+                    InterpreterAction::MachineCall(
+                        *id,
+                        arguments
+                            .iter()
+                            .map(|a| match a {
+                                MachineCallArgument::Unknown(v) => {
+                                    MachineCallArgumentIdx::Unknown(var_mapper.map_var(v))
+                                }
+                                MachineCallArgument::Known(e) => {
+                                    // convert known arguments into variable assignments that are then referenced
+                                    let idx = var_mapper.reserve_idx();
+                                    actions.push(InterpreterAction::AssignExpression(
+                                        idx,
+                                        var_mapper.map_expr_to_rpn(e),
+                                    ));
+                                    MachineCallArgumentIdx::Known(idx)
+                                }
+                            })
+                            .collect(),
+                    )
+                }
+                Effect::Branch(..) => {
+                    unimplemented!("Branches are not supported in the interpreter yet")
+                }
+            };
+            actions.push(action);
+        })
+    }
+
+    fn write_data(
+        var_mapper: &mut VariableMapper,
+        actions: &mut Vec<InterpreterAction<T>>,
+        effects: &[Effect<T, Variable>],
+    ) {
+        effects
+            .iter()
+            .flat_map(written_vars_in_effect)
+            .for_each(|(var, _mutable)| {
+                match var {
+                    Variable::Cell(cell) => {
+                        let idx = var_mapper.get_var(var).unwrap();
+                        actions.push(InterpreterAction::WriteCell(idx, cell.clone()));
+                        actions.push(InterpreterAction::WriteKnown(cell.clone()));
+                    }
+                    Variable::Param(i) => {
+                        let idx = var_mapper.get_var(var).unwrap();
+                        actions.push(InterpreterAction::WriteParam(idx, *i));
+                    }
+                    Variable::MachineCallReturnValue(_) => {
+                        // This is just an internal variable.
+                    }
+                }
+            });
+    }
+
+    // Check that actions are valid (e.g., variables writen to only once, and only read after being written to)
+    fn is_valid(&self) -> bool {
```

Comment:
Could be a function on a data structure that also performs conversion, contains the InterpreterActions, but not the code for actually interpreting.

Reply:
i'd do it in a separate PR (see my previous comment)